core-plugins/src/main/java/co/cask/hydrator/plugin/transform/JavaScriptTransform.java

```diff
@@ -0,0 +1,381 @@
+    super.initialize(context);
+    metrics = context.getMetrics();
+    logger = LoggerFactory.getLogger(ScriptTransform.class.getName() + " - Stage:" + context.getStageName());
+
+    // for Nashorn (Java 8+) support -- get method to convert ScriptObjectMirror to List
+    try {
+      Class<?> somClass = Class.forName("jdk.nashorn.api.scripting.ScriptObjectMirror");
+      somValuesMethod = somClass.getMethod("values");
+    } catch (NoSuchMethodException e) {
+      throw new RuntimeException(
+        "Failed to get method ScriptObjectMirror#values() for converting ScriptObjectMirror to List. " +
+          "Please check your version of Nashorn is supported.", e);
+    } catch (ClassNotFoundException e) {
+      // Ignore -- we don't have Nashorn, so no need to handle Nashorn
+    }
+
+    init(context);
+  }
+
+  @Override
+  public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter) {
+    try {
+      engine.eval(String.format("var %s = %s;", VARIABLE_NAME, GSON.toJson(input)));
+      Emitter<Map> jsEmitter = new JSEmitter(emitter, schema == null ? input.getSchema() : schema);
+      engine.put(EMITTER_NAME, jsEmitter);
+      invocable.invokeFunction(FUNCTION_NAME);
+    } catch (Exception e) {
+      throw new IllegalArgumentException("Could not transform input: " + e.getMessage(), e);
+    }
+  }
+
+  /**
+   * Emitter to be used from within Javascript code
+   */
+  public final class JSEmitter implements Emitter<Map> {
+
+    private final Emitter<StructuredRecord> emitter;
+    private final Schema schema;
+
+    public JSEmitter(Emitter<StructuredRecord> emitter, Schema schema) {
+      this.emitter = emitter;
+      this.schema = schema;
+    }
+
+    public void emit() {
+      System.out.println("Testing emit without params");
+    }
+
+    @Override
+    public void emit(Map value) {
+      emitter.emit(decode(value));
+    }
+
+    @Override
+    public void emitError(InvalidEntry<Map> invalidEntry) {
+      emitter.emitError(new InvalidEntry<>(invalidEntry.getErrorCode(), invalidEntry.getErrorMsg(),
+                                           decode(invalidEntry.getInvalidRecord())));
+    }
+
+    public void emitError(Map invalidEntry) {
+      emitter.emitError(getErrorObject(invalidEntry, decode((Map) invalidEntry.get("invalidRecord"))));
+    }
+
+    private StructuredRecord decode(Map nativeObject) {
+      return decodeRecord(nativeObject, schema);
+    }
+  }
+
+  private InvalidEntry<StructuredRecord> getErrorObject(Map result, StructuredRecord input) {
+    Preconditions.checkState(result.containsKey("errorCode"));
+
+    Object errorCode = result.get("errorCode");
+    Preconditions.checkState(errorCode instanceof Number,
+                             "errorCode entry in resultMap is not a valid number. " +
+                               "please check your script to make sure error-code is a number");
+    int errorCodeInt;
```

Comment:
javascript when it returns the error-code its type is double, so we can't directly cast it to int, that would fail, that's why we are having this additional step.